flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2304,7 +2304,10 @@ class ScalarExprLowering {
   genval(const Fortran::evaluate::FunctionRef<Fortran::evaluate::Type<TC, KIND>>
              &funRef) {
     auto retTy = converter.genType(TC, KIND);
-    return genProcedureRef(funRef, {retTy});
+    ExtValue result = genProcedureRef(funRef, {retTy});
+    if (result.rank() == 0 && fir::isa_ref_type(fir::getBase(result).getType()))
```

Comment:
Should this be more specific? Maybe loading only if the return value is a fir.ptr (or fir.heap)?

Reply:
I could indeed restrict this code more to the exact use case, but shouldn't all scalars be "loaded" by `genval` methods if they are in memory (I am putting "" around load because `genLoad` does not load objects that should always be handled in memory like characters) ? If so, I do not really see the point of making the test here more complex/narrow.

Reply:
Yes, the intention of `genval` should be to always return the value and not the address of the value. I don't remember if that design point was compromised in corner cases.